components/log-ingestor/src/ingestion_job/sqs_listener.rs

```diff
@@ -0,0 +1,230 @@
+    config: SqsListenerConfig,
+    sender: mpsc::Sender<ObjectMetadata>,
+}
+
+impl<SqsClientManager: AwsClientManagerType<Client> + 'static> SqsListenerTask<SqsClientManager> {
+    /// Runs the SQS listener task to listen to SQS messages and extract S3 object metadata. The
+    /// extracted metadata is sent to the provided channel sender.
+    ///
+    /// # Returns
+    ///
+    /// `Ok(())` on success.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    ///
+    /// * Forwards [`AwsClientManagerType::get`]'s return values on failure.
+    /// * Forwards [`Self::process_sqs_response`]'s return values on failure.
+    /// * Forwards
+    ///   [`aws_sdk_sqs::operation::receive_message::builders::ReceiveMessageFluentBuilder::send`]'s
+    ///   return values on failure.
+    pub async fn run(self, cancel_token: CancellationToken) -> Result<()> {
+        let mut polling_backoff_sec = self.config.init_polling_backoff_sec;
+        loop {
+            select! {
+                // Cancellation requested.
+                () = cancel_token.cancelled() => {
+                    return Ok(());
+                }
+
+                // Listen to SQS messages.
+                result = self.sqs_client_manager.get().await?
+                    .receive_message()
+                    .queue_url(self.config.queue_url.as_str())
+                    .max_number_of_messages(self.config.max_num_messages_to_fetch)
+                    .wait_time_seconds(polling_backoff_sec).send() => {
+                    polling_backoff_sec = if self.process_sqs_response(result?).await? {
+                        self.config.init_polling_backoff_sec
+                    } else { std::cmp::min(
+                        polling_backoff_sec * 2,
+                        self.config.max_polling_backoff_sec
+                    ) };
+                }
+            }
+        }
+    }
+
+    /// Processes the SQS response to extract S3 object metadata and send it to the channel sender.
+    ///
+    /// # NOTE
+    ///
+    /// The message will be deleted from the SQS queue if it is successfully processed.
+    ///
+    /// # Returns
+    ///
+    /// Whether any relevant object metadata was ingested.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    ///
+    /// * Forwards [`mpsc::Sender::send`]'s return values on failure.
+    /// * Forwards
+    ///   [`aws_sdk_sqs::operation::delete_message::builders::DeleteMessageFluentBuilder::send`]'s
+    ///   return values on failure.
+    async fn process_sqs_response(&self, response: ReceiveMessageOutput) -> Result<bool> {
+        if response.messages.is_none() {
+            return Ok(false);
+        }
+        let mut ingested = false;
+        for msg in response.messages.unwrap() {
+            if msg.body.is_none() {
+                continue;
+            }
+
+            let event = match serde_json::from_str::<S3>(msg.body.as_ref().unwrap()) {
```

Comment:
```suggestion
        let Some(messages) = response.messages else {
            return Ok(false);
        };
        let mut ingested = false;
        for msg in messages {
            let Some(body) = &msg.body else {
                continue;
            };

            let event = match serde_json::from_str::<S3>(body) {
```
how about more rusty